File: 2018/Case.Subs.KeyMatcher/Entry/CmdMain.cs
using System;
using Autodesk.Revit.Attributes;
using Autodesk.Revit.DB;
using Autodesk.Revit.UI;
using Case.Subs.KeyMatcher.Utility;

namespace Case.Subs.KeyMatcher.Entry
{

  [Transaction(TransactionMode.Manual)]
  public class CmdMain : IExternalCommand, IExternalCommandAvailability
  {

    /// <summary>
    /// Report Groups by View
    /// </summary>
    /// <param name="commandData"></param>
    /// <param name="message"></param>
    /// <param name="elements"></param>
    /// <returns></returns>
    public Result Execute(ExternalCommandData commandData, ref string message, ElementSet elements)
    {

      try
      {

        // Version
        if (!commandData.Application.Application.VersionName.Contains(clsGlobals.RevitYear))
        {
          clsGlobals.ShowTaskDialog(
            "Cannot Continue",
            "Revit Version Not Compatible",
            string.Format("This Add-In was built for Revit {0} and is not compatible with this version of Revit.\n\nPlease contact info@case-inc.com for assistance...", clsGlobals.RevitYear));
          return Result.Cancelled;
        }

        // Success
        return Result.Succeeded;

      }
      catch (Exception ex)
      {

        // Failure
        message = ex.Message;
        return Result.Failed;

      }

    }

    /// <summary>
    /// Command Availability
    /// </summary>
    /// <param name="applicationData"></param>
    /// <param name="selectedCategories"></param>
    /// <returns></returns>
    public bool IsCommandAvailable(UIApplication applicationData, CategorySet selectedCategories)
    {
      return applicationData.ActiveUIDocument != null;
    }

  }
}

File: 2018/Case.RoomInsertionPoint/Case.RoomInsertionPoint/Entry/CmdMain.vb
Imports Autodesk.Revit.Attributes
Imports Autodesk.Revit.DB
Imports Autodesk.Revit.UI
Imports System.Linq
Imports [Case].RoomInsertionPoint.API
Imports [Case].RoomInsertionPoint.Data

Namespace Entry

  <Transaction(TransactionMode.Manual)>
  Public Class CmdMain

    Implements IExternalCommand

    ''' <summary>
    ''' Command Entry Point
    ''' </summary>
    ''' <param name="commandData"></param>
    ''' <param name="message"></param>
    ''' <param name="elements"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function Execute(ByVal commandData As ExternalCommandData,
                            ByRef message As String,
                            ByVal elements As ElementSet) As Result Implements IExternalCommand.Execute

      Try

        ' Version
        If Not commandData.Application.Application.VersionName.Contains("2018") Then

          ' Failure
          Using td As New TaskDialog("Cannot Continue")
            With td
              .TitleAutoPrefix = False
              .MainInstruction = "Incompatible Version of Revit"
              .MainContent = "This Add-In was built, please contact CASE for assistance."
              .Show()
            End With
          End Using
          Return Result.Cancelled

        End If

        ' Rooms with Area > 0
        Dim m_listRooms As IEnumerable(Of Architecture.Room) = From e In New FilteredElementCollector(commandData.Application.ActiveUIDocument.Document) _
                                                               .OfCategory(BuiltInCategory.OST_Rooms)
                                                               Let r = TryCast(e, Architecture.Room)
                                                               Select r Where r.Area > 0

        ' Spaces with Area > 0
        Dim m_listSpaces As IEnumerable(Of Autodesk.Revit.DB.Mechanical.Space) = From e In New FilteredElementCollector(commandData.Application.ActiveUIDocument.Document) _
                                                                                 .OfCategory(BuiltInCategory.OST_MEPSpaces)
                                                                                 Let r = TryCast(e, Autodesk.Revit.DB.Mechanical.Space)
                                                                                 Select r Where r.Area > 0

        ' Total
        Dim iTotal As Integer = 0
        If Not m_listRooms Is Nothing Then iTotal += m_listRooms.Count
        If Not m_listSpaces Is Nothing Then iTotal += m_listSpaces.Count

        ' Progress
        Using p As New form_Progress(iTotal)
          p.Show()

          ' Process Each Room
          If Not m_listRooms Is Nothing Then
            For Each x In m_listRooms.ToList

              Try
                p.StepProgress()
              Catch
              End Try

              ' Helper
              Dim m_e As New clsRoom(x)
              m_e.MoveToCentroid()

            Next
          End If

          ' Process Each Space
          If Not m_listSpaces Is Nothing Then
            For Each x In m_listSpaces.ToList

              Try
                p.StepProgress()
              Catch
              End Try

              ' Helper
              Dim m_e As New clsRoom(x)
              m_e.MoveToCentroid()

            Next
          End If

          ' Progress
          p.Close()

        End Using

        ' Success
        Return Result.Succeeded

      Catch ex As Exception

        ' Failure
        message = ex.Message
        Return Result.Failed

      End Try

    End Function

  End Class
End Namespace